sql/temporary_tables.cc

```diff
@@ -671,7 +671,12 @@ bool THD::drop_temporary_table(TABLE *table, bool *is_trans, bool delete_table)
   temporary_tables->remove(share);
 
   /* Free the TABLE_SHARE and/or delete the files. */
-  free_tmp_table_share(share, delete_table);
+  if (free_tmp_table_share(share, delete_table))
+  {
```

Comment:
This should set `result= true`. which addresses the "caller doesn't know if ha_delete_table() failed or not" (from the MDEV).

The sql_print_warning here is redundant as `THD::rm_temporary_table` already does the warning so can be omitted.

In 10.5 `THD::rm_temporary_table`
```
b73d85277939 (Daniel Black        2022-03-16 17:51:49 +1100  712)   if (mysql_file_delete(key_file_frm, frm_path,
b73d85277939 (Daniel Black        2022-03-16 17:51:49 +1100  713)                         MYF(MY_WME | MY_IGNORE_ENOENT)))
bfed2c7d57a7 (Sergei Golubchik    2022-03-11 20:18:22 +0100  714)     error= true;
b73d85277939 (Daniel Black        2022-03-16 17:51:49 +1100  715) 
```
Warning was emitted on delete (added in 5bcb1d65328effd570031ae43bda78c3da4b9a6f (10.5)), except for missing.

10.4 looks like:
```
bfed2c7d57a7 (Sergei Golubchik    2022-03-11 20:18:22 +0100  706)   if (mysql_file_delete(key_file_frm, frm_path, MYF(0)))
bfed2c7d57a7 (Sergei Golubchik    2022-03-11 20:18:22 +0100  707)   {
bfed2c7d57a7 (Sergei Golubchik    2022-03-11 20:18:22 +0100  708)     error= true;
bfed2c7d57a7 (Sergei Golubchik    2022-03-11 20:18:22 +0100  709)   }
```

To avoid changing the 10.4 (needing to backport the `MY_IGNORE_ENOENT`), can you rebase on to 10.5 as that won't require a `THD::rm_temporary_table` change.

Can a test case in the form of the following work:
```
--source include/not_as_root.inc
CREATE TEMPORARY TABLE t (..);

(if the file doesn't exist - skip-test  - my fault - see 3edac3f18d2d74f68e6f9f9b41e5d05a5c8ca9c5)
--chmod  0000 {the temporary file}
--error {something}
DROP TEMPORARY TABLE t;
```

To get the test case working you may need to temporary hack:
```
--- a/mysys/mf_tempfile.c
+++ b/mysys/mf_tempfile.c
@@ -114,7 +114,7 @@ File create_temp_file(char *to, const char *dir, const char *prefix,
     dir= DEFAULT_TMPDIR;
 #ifdef O_TMPFILE
   {
-    static int O_TMPFILE_works= 1;
+    static int O_TMPFILE_works= 0;
 
     if ((MyFlags & MY_TEMPORARY) && O_TMPFILE_works)
     {

```